### PR TITLE
SQLite Migrations: Rebuild tables

### LINKED ISSUE #329: Sqlite Migrations: Table rebuilds
In SQLite, the following operations will require rebuilding the table - https://docs.microsoft.com/en-us/ef/core/providers/sqlite/limitations.

Other databases also have a smaller set of limitations that require a table rebuild (i.e. changing Identity on a column in SQL Server).

### PR DESCRIPTION
**Pull request by bricelam**
More tests to come, but I think we can get something merged and iterate on it.

Fixes #329

### CODE REVIEW

File: test/EFCore.Sqlite.FunctionalTests/MigrationsSqliteTest.cs [hunk 1]
```diff
@@ -278,139 +276,330 @@ public override async Task Add_column_with_max_length()
...
+    ""Id"" INTEGER NOT NULL
+);",
+                @"INSERT INTO ""ef_temp_People"" (""Id"")
+SELECT ""Id""
+FROM People;",
+                @"PRAGMA foreign_keys = 0;",
+                @"DROP TABLE ""People"";",
+                @"ALTER TABLE ""ef_temp_People"" RENAME TO ""People"";",
+                @"PRAGMA foreign_keys = 1;");
+        }
```
Comment (bricelam) [inline test/EFCore.Sqlite.FunctionalTests/MigrationsSqliteTest.cs:561]: 😍😍😍

File: src/EFCore.Sqlite.Core/Migrations/SqliteMigrationsSqlGenerator.cs [hunk 2]
```diff
@@ -67,44 +68,459 @@ private bool IsSpatialiteColumn(AddColumnOperation operation, IModel model)
...
+                        {
+                            var overrides = rebuild.NewlyRequiredColumns.GetOrAddNew(alterColumnOperation.Name);
+                            overrides.DefaultValueSql = alterColumnOperation.DefaultValueSql;
+                            overrides.DefaultValue = alterColumnOperation.DefaultValue;
+                        }
+
+                        break;
+                    }
+
+
```
Comment (smitpatel) [inline src/EFCore.Sqlite.Core/Migrations/SqliteMigrationsSqlGenerator.cs:124]:
```suggestion

```

**Comment by roji** [comment]
Just a thought... Should we maybe think about a migration warning when we do a table (or even column) rebuild, similar to how we have a warning for potential data loss? Where lots of data is involved, rebuilds can be pretty huge operations that users should execute with great care. Even an index rebuild (am thinking about the ALTER COLUMN for collation changes, for example) can be quite a big operation.

(not urgent, can be introduced later, etc...)

**Comment by bricelam** [comment]
Design meeting notes:
* Verify what happens with self-referencing foreign key constraints
* We only need to create indexes before the INSERT if they're new
  * Existing UNIQUE indexes already ensure that the data is valid
  * We can create them under their actual name since they won't conflict

**Pull request merged**